blockchain/chain.go

```diff
@@ -137,17 +122,49 @@ func removeChildNode(children []*blockNode, node *blockNode) []*blockNode {
 	return children
 }
 
+// BestState houses information about the current best block and other info
+// related to the state of the main chain as it exists from the point of view of
+// the current best block.
+//
+// The BestSnapshot method can be used to obtain access to this information
+// in a concurrent safe manner and the data will not be changed out from under
+// the caller when chain state changes occur as the function name implies.
+// However, the returned snapshot must be treated as immutable since it is
```

Comment:
Makes me wish Go had support for compiler enforced immutable structs. Which might look something like this: 
```go
type BestState immutable struct {
 ...
}
```

However, it's possible to emulate this sort of struct immutability today. We can do this by moving `BestState` to its own module and making all the attributes of the struct private (lower case), and only providing access of the attributes via publicly exported methods.